processor/signoztailsampler/evaluator.go

```diff
@@ -0,0 +1,141 @@
+	// if any of subpolicy filters match, the sampling method associated with that
+	// sub-policy will be applied. and no further processing will be performed.
+	subpolicies []sampling.PolicyEvaluator
+	logger      *zap.Logger
+}
+
+func NewDefaultEvaluator(logger *zap.Logger, policyCfg *PolicyCfg) sampling.PolicyEvaluator {
+
+	// condition operator to apply on filters (AND | OR)
+	filterOperator := policyCfg.PolicyFilterCfg.FilterOp
+
+	// list of filter evaluators used to decide if this policy
+	// should be applied
+	filters := prepareFilterEvaluators(logger, policyCfg.PolicyFilterCfg)
+	// todo(amol): need to handle situations with zero filters
+
+	// list of sub-policies evaluators
+	subpolicies := make([]sampling.PolicyEvaluator, len(policyCfg.SubPolicies))
+	for _, subRule := range policyCfg.SubPolicies {
+		subPolicy := NewDefaultEvaluator(logger, &subRule)
+		subpolicies = append(subpolicies, subPolicy)
+	}
+
+	// sampling is applied only when filter conditions are met
+	var sampler sampling.PolicyEvaluator
+	var samplingMethod string
+	switch policyCfg.SamplingPercentage {
+	case 0:
+		samplingMethod = "exclude_all"
+		sampler = sampling.NewAlwaysUnsample(logger)
+	case 100:
+		samplingMethod = "include_all"
+		sampler = sampling.NewAlwaysSample(logger)
+	default:
+		samplingMethod = "probabilistic"
+		sampler = sampling.NewProbabilisticSampler(logger, "21d2", policyCfg.SamplingPercentage)
+	}
+
+	return &defaultEvaluator{
+		name:           policyCfg.Name,
+		root:           policyCfg.Root,
+		sampler:        sampler,
+		samplingMethod: samplingMethod,
+		filterOperator: filterOperator,
+		filters:        filters,
+		subpolicies:    subpolicies,
+	}
+}
+
+func prepareFilterEvaluators(logger *zap.Logger, policyFilterCfg PolicyFilterCfg) []sampling.PolicyEvaluator {
+	var filterEvaluators []sampling.PolicyEvaluator
+	for _, s := range policyFilterCfg.StringAttributeCfgs {
+		filterEvaluators = append(filterEvaluators, sampling.NewStringAttributeFilter(logger, s.Key, s.Values, s.EnabledRegexMatching, s.CacheMaxSize, s.InvertMatch))
+	}
+
+	for _, n := range policyFilterCfg.NumericAttributeCfgs {
+		filterEvaluators = append(filterEvaluators, sampling.NewNumericAttributeFilter(logger, n.Key, n.MinValue, n.MaxValue))
+	}
+	// todo: status filter
+	return filterEvaluators
+}
+
+func (de *defaultEvaluator) Evaluate(traceId pcommon.TraceID, trace *sampling.TraceData) (sampling.Decision, error) {
+
+	// todo: explore doing this in parallel
+	for _, sp := range de.subpolicies {
+		decision, err := sp.Evaluate(traceId, trace)
+		if err != nil {
+			// todo: consider adding health for each evaluator
+			// to avoid printing log messages for each trace
+			zap.S().Errorf("failed to evaluate trace:", traceId)
+			continue
+		}
+
+		if decision != sampling.NoResult {
+			// found a result, exit
+			return decision, nil
+		}
+	}
+
+	// filterMatched captures when atleast one filter matches
+	var filterMatched bool
+
+	for _, fe := range de.filters {
```

Comment:
It says `policy_filters` is what activates the policy, but here it matches after the sub-policies? Say I have a service where I want to do 20% probabilistic on key `my-custom-key-1` and 80% on ``my-custom-key-2` then my policy would be
```
{
	Name:     "test-policy-1",
	Type:     PolicyGroup,
	Priority: 1,
	Root:     true,
	PolicyFilterCfg: PolicyFilterCfg{
		StringAttributeCfgs: []StringAttributeCfg{
			{Key: "service.name", Values: []string{"route"}},
		},
	},
	SubPolicies: []PolicyCfg{
		{
			Name:     "custom-condition-1",
			Type:     PolicyGroup,
			Priority: 1,
			ProbabilisticCfg: ProbabilisticCfg{
				SamplingPercentage: 20,
			}
			PolicyFilterCfg: PolicyFilterCfg{
				StringAttributeCfgs: []StringAttributeCfg{
					{Key: "my-custom-key-1", Values: []string{"some-value-1"}},
				},
			},
			{
				Name:     "custom-condition-2",
				Type:     PolicyGroup,
				Priority: 1,
				ProbabilisticCfg: ProbabilisticCfg{
					SamplingPercentage: 80,
				}
				PolicyFilterCfg: PolicyFilterCfg{
					StringAttributeCfgs: []StringAttributeCfg{
						{Key: "my-custom-key-2", Values: []string{"some-value-2"}},
					},
				},
			},
		},
	},
},
```

Why does the evaluator go to sub-policies without making sure my service matches? Or did I misunderstand something? I don